### PR TITLE
Add cache

### PR DESCRIPTION
**Pull request by TobleMiner**
This PR adds caching support to mesh-announce.

There have been a few changes to support caching:
* All code snippets are now fully featured python modules:
  - Each module provides a class instance which represents a part of a respondd provider
  - Fine grained cache control is easily possible by per module ttl setting

As a direct consequence of treating providers as a set of modules all providers are now only compiled once.

This PR increases the performance of mesh-announce drastically. Once the cache is populated all further requests do only need around 3 ms of processing time. Compared to the previous response time of about 120 ms to ```nodeinfo``` requests this provides a 4000 % speedup.

### CODE REVIEW

**Comment by rubo77** [comment]
if there is no other meaning within one week, I would merge this.

**Comment by TobleMiner** [comment]
Nope, there wasn't. I wanted fine grained control over cache TTLs and tried to eliminate ugly manual code compilation as well as any hacky environment modification to get shorthands within code snippets.  Thus I rewrote the modules, too.

In retrospect I should have probably splitted this into multiple commits but I just didn't have the patience to do so back then.

File: providers/__init__.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    if len(path) == 1:
+        node[key] = value
+    elif key in node:
+        _set_value(node[key], path[1:], value)
+    else:
+        node[path[0]] = {}
+        _set_value(node[key], path[1:], value)
+
+
+def _compile_file(fname):
```
[the remark this answers is not in the document]
Comment (TobleMiner) [reply providers/__init__.py]: Good catch. I'll remove that

File: providers/__init__.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+class DataSource():
+    ''' Base data source, inherited by sources in provider directories
+    '''
+    def cache_ttl(self):
+        ''' Default cache ttl, set to 0 to disable cache
+        '''
+        return 30
+
+    def required_args(self):
```
[the remark this answers is not in the document]
Comment (TobleMiner) [reply providers/__init__.py]: I think my solution is a little cleaner. First of all we can check if all required arguments are actually available before calling the cal function. Additionally we don't need to extract the argument from a dict inside the module

File: providers/__init__.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
+import os
+import time
+import traceback
+from util import source_dirs, find_modules
+from importlib import import_module
+
+class DataSource():
+    ''' Base data source, inherited by sources in provider directories
+    '''
+    def cache_ttl(self):
```
[the remark this answers is not in the document]
Comment (TobleMiner) [reply providers/__init__.py]: I think using an attribute here makes implementing new data sources less prone to errors. If I was to implement the ttl as an attribute every module would have to implement a constructor including an explicit super().__init__() call (assuming the parent constructor may do something useful in the future) to change the ttl. Thus I do favor a function over an attribute

File: util.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+def modules(dir):
+    return [os.path.splitext(f)[0] for f in next(os.walk(dir))[2]
+        if f.endswith('.py') and not f.startswith('__')]
+
+def find_modules(base, *path):
+    path_files = []
+    dir_abs = base
+    for dir in path:
+        dir_abs = os.path.join(dir_abs, dir)
```
[the remark this answers is not in the document]
Comment (TobleMiner) [reply util.py]: Yep, didn't realize `path.join` can handle more than two arguments

File: util.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import os
+
+def source_dirs(dir):
+    dirs = next(os.walk(dir))[1]
+    return filter(lambda d: not d.startswith('__'), dirs)
```
[the remark this answers is not in the document]
Comment (TobleMiner) [reply util.py]: Sounds like a good idea. I'll implement that

**Comment by TobleMiner** [comment]
@jplitza Hi, sorry for my lack of response. I got carried away doing a bunch of other stuff.

I've updated this PR to implement most of your suggested changes.

**Comment by rubo77** [comment]
instead of accepting this PR, we can also accept the PR #33, which is includiong this PR

**Comment by rubo77** [comment]
was included in #33

**Pull request closed without merging**